internal/graph/track_check_resolver.go

```diff
@@ -0,0 +1,208 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+const (
+	trackerLogLines    = 10
+	trackerLogBurst    = 15
+	trackerLogInterval = time.Second
+	trackerInterval    = time.Duration(60) * time.Second
+)
+
+// TrackerCheckResolverOpt defines an option pattern that can be used to change the behavior of TrackerCheckResolver.
+type TrackerCheckResolverOpt func(checkResolver *TrackerCheckResolver)
+type resolutionTree struct {
+	tm   time.Time
+	hits *atomic.Uint64
+}
+
+type TrackerCheckResolver struct {
+	delegate CheckResolver
+	ticker   *time.Ticker
+	logger   logger.Logger
+	limiter  *rate.Limiter
+	ctx      context.Context
+	nodes    sync.Map
+}
+
+var _ CheckResolver = (*TrackerCheckResolver)(nil)
+
+func WithTrackerLogger(logger logger.Logger) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.logger = logger
+	}
+}
+
+func WithTrackerContext(ctx context.Context) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.ctx = ctx
+	}
+}
+
+// Expired check the current tuple entry expiration.
+func (r *resolutionTree) expired() bool {
+	return time.Since(r.tm) > trackerInterval
+}
+
+// NewTrackCheckResolver creates an instance tracker Resolver.
+func NewTrackCheckResolver(opts ...TrackerCheckResolverOpt) *TrackerCheckResolver {
+	t := &TrackerCheckResolver{
+		limiter: rate.NewLimiter(rate.Limit(trackerLogLines/trackerLogInterval), trackerLogBurst),
+		ticker:  time.NewTicker(trackerInterval),
+	}
+
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	t.delegate = t
+	t.launchFlush()
+	return t
+}
+
+// LogExecutionPaths reports the model and tuple path.
+func (t *TrackerCheckResolver) logExecutionPaths(flush bool) {
+	t.nodes.Range(func(k, v any) bool {
+		modelid := k.(string)
+		paths, _ := v.(*sync.Map)
+		paths.Range(func(k, v any) bool {
+			tree, _ := v.(*resolutionTree)
+			path := k.(string)
+			if tree.expired() || flush {
+				if t.limiter.Allow() {
+					t.logger.Info("execution path hits",
+						zap.String("model", modelid),
+						zap.String("path", path),
+						zap.Uint64("hits", tree.hits.Load()))
+					paths.Delete(path)
```

Comment:
If a request has many "paths" this will generate a crap ton of logs

Reply:
The rate limiter controls the times when the logs will be emmited, but not the number of logs emmited, right?

Reply:
That's correct.